python/tests/test_search_algorithm.py

```diff
@@ -1,73 +1,97 @@
 from dataclasses import dataclass
-from typing import Any, Sequence
+from typing import Sequence, Mapping, Dict
 from unittest import TestCase
 
-from src.search_algorithm import CandidateClass, CandidateGenerator, \
-    CostFunction, CostMap, SearchAlgorithm
+from src.search_algorithm import (
+    CandidateClass,
+    CandidateGenerator,
+    CostMap,
+    SearchAlgorithm,
+)
 
 
 @dataclass(frozen=True)
 class TestCandidate(CandidateClass):
-    ckey: str = ""
-    cost: float = 0.0
+    name: str = ""
+    value: float = 0.0
 
     def key(self) -> str:
-        return self.ckey
+        return self.name
 
 
 class TestCandidateGenerator(CandidateGenerator[TestCandidate]):
     # pylint: disable=no-self-use
+    def __init__(self, initial_population: Sequence[float]):
+        self._initial_population = initial_population
+
     def get_initial_population(self) -> Sequence[TestCandidate]:
-        return tuple(TestCandidate(f"cand_{i}", float(i)) for i in range(5))
+        return tuple(
+            TestCandidate(f"cand_{i}", val)
+            for i, val in enumerate(self._initial_population)
+        )
 
     def get_next_population(
-            self,
-            current_candidates: Sequence[TestCandidate],
-            cost_map: CostMap
+        self, current_candidates: Sequence[TestCandidate], cost_map: CostMap
     ) -> Sequence[TestCandidate]:
-        expected_cost_map = {candidate.ckey: candidate.cost for candidate in current_candidates}
-        assert cost_map == expected_cost_map
-        return current_candidates
+        return tuple(
+            candidate
+            for candidate in current_candidates
+            if cost_map[candidate.key()] == 0.0
+        )
+
 
+def test_cost_function(
+    candidates: Sequence[TestCandidate],
+    simulation_result: Mapping[str, float],
+) -> CostMap:
+    return {
+        candidate.key(): simulation_result[candidate.key()] for candidate in candidates
+    }
 
-class TestCostFunction(CostFunction[TestCandidate, Any]):
-    # pylint: disable=no-self-use,unused-argument
-    def calculate(self, candidates: Sequence[TestCandidate], simulation_result: Any) -> CostMap:
-        return {candidate.ckey: candidate.cost for candidate in candidates}
 
+def test_simulator(candidates: Sequence[TestCandidate]) -> Dict[str, float]:
+    result = {candidate.key(): 0.0 for candidate in candidates}
+    if candidates[0].value < candidates[1].value:
+        result[candidates[0].key()] = 1.0
+    else:
+        result[candidates[1].key()] = 1.0
+    return result
```

Comment:
Not sure I understand the point of this. Generally you want test logic to be as simple as possible so you're just testing the production code, not adding too much functionality to the test method. I think we should just hardcode the result, maybe `value == index`?

Reply:
I want a test for the search algorithm interface that is as simple as possible yes, but which actually tests your search algorithm code. This is the simplest search algorithm I could come up with that actually uses the same structure. It is a contrived algorithm for the purpose of the test. 

The way the algorithm works it wouldn't test anything if I just used the value or index as the result, because I want a test that takes multiple steps to converge so I could test the convergence detector. I also wanted to test the candidate generator. 

I think the most proper way would be to include individual unit tests for these mock functions, and then use them for the integration test of the algorithm, but I think they are simple enough to trust without that without adding too much risk given our time constraints

Reply:
Ah okay, I see you upped the iteration count to do multiple round. Sounds good 👍